plutus-ir/src/Language/PlutusIR/Transform/LetFloat.hs

```diff
@@ -0,0 +1,455 @@
+      in M.delete lamName -- this trick removes any lamrank from the final value, since we don't care about lambdas in the final pass1 result (P1Data)
+         <$> (local (\ (scope,oldDepth) ->
+                        let newDepth = oldDepth+1
+                            newRank = LamDep newDepth lamName
+                            newScope = M.insert lamName (newRank,lamName) scope -- adds a new lam rank to the current scope
+                        in (newScope, newDepth))
+                    $ goTerm tBody)
+
+    goRec :: [Binding tyname name uni a] -> Term tyname name uni a -> Reader P1Ctx P1Data
+    goRec bs tIn = do
+        -- the freevars is the union of each binding's freeVars,
+        -- excluding the newly-introduced binding identifiers of this letrec
+        let freeVars = foldMap fBinding bs S.\\ foldMap bindingIds bs
+        -- all bindings share their commonly-maximum rank
+        newScope <- addRanks bs freeVars
+        resBs <- mconcat <$> forM (bs^..traverse.bindingSubterms)
+                                  (withScope newScope . goTerm)
+        resIn <- withScope newScope $ goTerm tIn
+        pure $ resBs <> resIn
+
+    goNonRec :: [Binding tyname name uni a] -> Term tyname name uni a -> Reader P1Ctx P1Data
+    goNonRec bs tIn =
+      foldr (\ b acc -> do
+           -- this means that we see each binding individually, not at the whole let level
+           let freeVars = fBinding b
+           -- compute a maxrank for this binding alone and a newscope that includes it
+           newScope <- addRanks [b] freeVars
+           resHead <- mconcat <$> forM (b^..bindingSubterms) goTerm
+           resRest <- withScope newScope acc
+           pure $ resHead <> resRest
+       ) (goTerm tIn) bs
+
+
+    -- | Given a set of bindings and the union set of their free-variables,
+    -- compute the maximum 'Rank' among all the free-variables,
+    -- and increase the current scope ('P1Data') by inserting mappings of each new let-identifier to this maximum rank
+    addRanks :: [Binding tyname name uni a] -- ^ bindings
+             -> S.Set PLC.Unique -- ^ their free-vars
+             -> Reader P1Ctx P1Data -- ^ the updated scope that includes the added ranks
+    addRanks bs freeVars = do
+      (scope, _) <- ask
+      let
+          -- the ranks of the free variables
+          freeRanks = fmap fst $ M.restrictKeys scope freeVars
+          -- the max rank from all their free variables
+          maxRank = fold freeRanks
+          -- this computed max rank is linked/associated to all the given bindings and added to the current scope
+          newScope = foldr (\ b acc ->
+                               -- this fold is needed because a datatype-binding may introduce multiple identifiers
+                               -- and we need to create scope entries for each one of them, having the same maxrank
+                               foldr (\i ->  M.insert i (maxRank, b^.principal)) acc $ bindingIds b
```

Comment:
You can make this faster and more readable by using [`fromList`](http://hackage.haskell.org/package/containers-0.6.1.1/docs/Data-Map-Strict.html#v:fromList).

Reply:
Ah, it'the initial value is `acc`, not Data.Map.empty`. Anyway, I'd make this explicit: there is an original `scope`, you create `scopeDelta` by processing all the `bindingIds` of each of the binding and then I'd define ``newScope = scope `M.union` scopeDelta``.

I think this would read better, but no big deal.

Reply:
Turned it into a list comprehension. hope ppl still use those.